### ISSUE TITLE
after installing, libratbag, ratbagd and piper, when I run piper nothing happens

### ISSUE THREAD

**Issue by Chronigan** [libratbag/piper#15]
What information would you like me to give you for this issue?

**Comment by Hjdskes** [comment]
First of all, thanks for your persistence in the matter!

I'm not sure what's going wrong on your end; I just installed everything from current git myself and it works for me. Since there are no installation instructions for Piper (yet); how did you install it? The process is the same for libratbag and ratbagd (i.e. `meson builddir --prefix=/usr`, `ninja -C build` and `sudo ninja -C build install`).

A few questions:

1. Is anything printed to the terminal when you run Piper from there?
2. Is ratbagd running? Make sure that it is using `systemctl start ratbagd`.
3. Is the mouse that you're trying to configure supported by libratbag?

For 2 and 3, *something* should be displayed to you...

**Comment by Chronigan** [comment]
No problem, thank you for all your hard work!

When I run piper from the terminal, I get this error : 

```
richard@Mercury ~/Programs/piper/builddir $ piper
Traceback (most recent call last):
  File "/usr/bin/piper", line 9, in <module>
    from piper import piper
ImportError: No module named 'piper'

```
ratbagd is running and I have a logitech G602 mouse. I am running Linux Mint 18.1 and installed it by following the instructions for the other libraries.

**Comment by Hjdskes** [comment]
Hmm.. @whot also got this error when trying to install Piper. I'll add it to the top of my list and will get back to you.

**Comment by whot** [comment]
You need the PYTHONPATH to point to the srcdir, that looks like the main issue here. It's a bit sub-optimal when running from mesons' builddir because PYTHONPATH has to point to the sourcdir. We should probably add some code to detect when run from the builddir and add the path accordingly.

**Comment by bentiss** [comment]
A thing I had to do in the past: https://github.com/bentiss/hid-replay/blob/master/tools/parse_hut.py#L26

This fetches the actual path from the current file and allows to fetch data when not executed in the src directory. Maybe we can inspire from this, even if we might want the other way around.

**Comment by Hjdskes** [comment]
@bentiss  There already is code to detect running from source in [`piper.in`](https://github.com/libratbag/piper/blob/master/piper.in#L17). A larger issue here is that we need a built `piper.gresource` file, and we cannot enforce its creation at runtime (not without calling `glib-compile-resources` from code, which is ugly imo). But this issue is not about running Piper from the source directory.

**Comment by whot** [comment]
can we get the ninja install logs please and the output of 

      python3 -c "import sys; print('\n'.join(sys.path))"

**Comment by Chronigan** [comment]
ninja install log
```
# List of files installed by Meson
# Does not contain files installed by custom scripts.
/usr/lib/python3.5/site-packages/piper/ratbagd.py
/usr/lib/python3.5/site-packages/piper/__init__.py
/usr/lib/python3.5/site-packages/piper/piper.py
/usr/share/piper/piper.gresource
/usr/share/applications/piper.desktop
/usr/share/icons/hicolor/scalable/apps/piper.svg
/usr/bin/piper
```
```

richard@Mercury ~ $ python3 -c "import sys; print('\n'.join(sys.path))"

/usr/lib/python35.zip
/usr/lib/python3.5
/usr/lib/python3.5/plat-x86_64-linux-gnu
/usr/lib/python3.5/lib-dynload
/home/richard/.local/lib/python3.5/site-packages
/usr/local/lib/python3.5/dist-packages
/usr/lib/python3/dist-packages
```

```
```

**Comment by whot** [comment]
that's the problem here, /usr/lib/python3.5/site-packages/ where we install to isn't in your python path. I'm not sure if that's a meson error or some issue with your setup tbh. meson gets that value from Python's sysconfig module [1] so I'm guessing that any difference between the two is a Python configuration issue.

[1] https://docs.python.org/3/library/sysconfig.html

**Comment by Hjdskes** [comment]
Looks like this isn't a bug on our side, but I'm leaving it open a little while longer for @Chronigan to reply.

**Comment by Chronigan** [comment]
I don't understand. How should I change my python configuration? I have not altered it and this is a fresh install of Mint.

**Comment by whot** [comment]
What's the output of `python3 -c "import sysconfig; print(sysconfig.get_path('purelib'));"`?

**Comment by Chronigan** [comment]
`/usr/lib/python3.5/site-packages`

**Comment by Hjdskes** [comment]
Did you alter your `PYTHONPATH` environment variable? What does `echo $PYTHONPATH` say?

Piper should work for you if you run `PYTHONPATH=$PYTHONPATH:/usr/lib/python3.5/site-packages piper`, can you try that?

**Comment by Chronigan** [comment]
When I try running piper as you suggest, I get the following

```
Traceback (most recent call last):
  File "/usr/bin/piper", line 9, in <module>
    from piper import piper
  File "/usr/lib/python3.5/site-packages/piper/piper.py", line 4, in <module>
    from ratbagd import *
ImportError: No module named 'ratbagd'
```
I have installed ratbagd per the instructions on that page

echo $PYTHONPATH returns nothing

**Comment by Hjdskes** [comment]
The ratbagd error is as expected, since it recently dropped its bindings. They've been imported into Piper, but the old Piper code hasn't been updated. I'll fix this tomorrow.

Still not sure what's wrong with your setup, though...

**Comment by Chronigan** [comment]
I know it would be a lot of work but do you think you could set up a vm with mint 18.1 and try to get piper running? It might be something to do with mint or the version of the dependencies you can install on it from the synaptic package manager.

**Comment by Chronigan** [comment]
I really want to help ya'll get this out there as i think something like this should be shipped as default with distros.

**Comment by whot** [comment]
I'm almost 100% sure this is an issue somewhere within your python install.

From the official sysconfig docs: https://docs.python.org/3/library/sysconfig.html
> purelib: directory for site-specific, non-platform-specific files.
and `/usr/lib/python3.5/site-packages` seems like the correct output for python 3.5

From the official module docs: https://docs.python.org/3/tutorial/modules.html
> If not found, it then searches for a file named spam.py in a list of directories given by the variable sys.path

In your case, `sys.path` does not contain the path that's provided to meson by `sysconfig`, which explains the module load error. But if these aren't in sync, problems are expected. This isn't a problem with piper, we're using the path python tells us to use, piper just triggers this bug because it's using sysconfig through meson.

**Comment by Hjdskes** [comment]
@Chronigan I'll play around with it in a VM to see if I can figure something out, but it'll take a few days at least. I have other things to do with a higher priority first.

**Comment by whot** [comment]
This one should be fixed after #23, closing for now. Please re-open it if it's still an issue.

**Comment by ITCMD** [comment]
I'm having this same issue on KDE Neon on the latest build.

**Comment by whot** [comment]
@ITCMD please open a new issue, probably a new bug you're seeing. Attach the output from `ratbagctl list` and `ratbagctl info <device name>` as well as any output from piper when started from the terminal.

**Comment by TAOXUY** [comment]
Second to this issue

**Comment by TAOXUY** [comment]
BTW, is debian pkg supplied? Build from source really sucks.

**Comment by z3t4m4n** [comment]
I was having this issue with ubuntu 19.10, with the ppa; Then I upgraded to 20.04, sitll happening with both ppa & repo package, so I removed piper,
Noticed that still had the executable, investigated a bit and sure it was already installed with pip,  so I uninstalled it.
Checked again, the executable still was there, seems that time ago I built and installed from tarbal.
I found the piper executable with 'which', removed it, installed again from the oficial ubuntu repos and piper worked flawlessly.
